File: backend/modules/code/locus.py
"""Backend half of the **code-locus bus** — the shared "what code am I looking at"
cursor (path + optional range/symbol) every coding pane publishes to and follows.

Browsers mirror their locus here over the `code` `/ws` channel so `dash.code` and
the agent can read the live locus; the backend (dash) can set it and browsers
follow. Mirrors the library broadcaster's recorder/stream split (push-only fan-out;
the shared `/ws` receive loop owns inbound reads). Self-echo is suppressed on the
frontend by an `origin` client id — the backend just stores and re-fans. See
docs/modules/code.mdx.
"""

from __future__ import annotations

import asyncio
from typing import Any

# The latest locus any browser (or dash) reported. Read by `dash.code.locus()`.
_current_locus: dict[str, Any] = {}


class CodeBroadcaster:
    def __init__(self) -> None:
        self._subscribers: set[asyncio.Queue[dict[str, Any]]] = set()

    def subscribe(self) -> asyncio.Queue[dict[str, Any]]:
        queue: asyncio.Queue[dict[str, Any]] = asyncio.Queue(maxsize=100)
        self._subscribers.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict[str, Any]]) -> None:
        self._subscribers.discard(queue)

    def publish(self, event: dict[str, Any]) -> None:
        for queue in list(self._subscribers):
            try:
                queue.put_nowait(event)
            except asyncio.QueueFull:
                pass


code_events = CodeBroadcaster()


def current_locus() -> dict[str, Any]:
    """A copy of the most recent locus (empty dict if none reported yet)."""
    return dict(_current_locus)


def set_locus_from_backend(locus: dict[str, Any]) -> dict[str, Any]:
    """Set the locus from the backend (dash/agent) and push it to browsers to follow.
    Tagged `source='dash'` with no client `origin`, so every browser applies it."""
    locus = {**locus, "source": locus.get("source", "dash")}
    _current_locus.clear()
    _current_locus.update(locus)
    code_events.publish({"event": "locus", "data": locus})
    return locus


async def handle_code_message(conn: Any, msg: dict[str, Any]) -> None:
    """Inbound `code` channel: a browser published its locus. Store it (so dash and
    the agent can read the live locus) and re-fan to browsers so panes stay in sync
    across windows. The frontend `origin` tag lets the sender ignore its own echo."""
    if msg.get("event") != "locus":
        return
    data = msg.get("data")
    if not isinstance(data, dict):
        return
    _current_locus.clear()
    _current_locus.update(data)
    code_events.publish({"event": "locus", "data": data})


async def push_code_events(conn: Any) -> None:
    queue = code_events.subscribe()
    try:
        while True:
            event = await queue.get()
            await conn.send_json(
                {"channel": "code", "event": event["event"], "data": event["data"]}
            )
    finally:
        code_events.unsubscribe(queue)

File: packages/core/src/locus.ts
/**
 * The **code-locus bus** — the emacs `point`, for code. A single, global, ephemeral
 * "what code am I looking at": `{ path, range?, symbol? }`. Every coding pane
 * publishes to it (`setLocus`) and follows it (`useLocus`), so one click in the
 * outline drives the editor, and the editor's cursor drives the outline — without any
 * pane knowing about another.
 *
 * A core primitive (not a module), because it's the shared spine multiple modules
 * hang off. Updates mirror over the `code` `/ws` channel so `dash.code`, the agent,
 * and other windows follow. Two guards keep it from feeding back on itself (the same
 * hazard the reveal/toggle buses hit): a per-client `origin` id drops our own echoes
 * off the socket, and each update carries the `source` pane so a follower can ignore
 * updates it drove itself. See docs/modules/code.mdx.
 */
import { useSyncExternalStore } from 'react';

import { subscribeChannel, sendChannel, type WsMessage } from './ws';

export interface LocusPosition {
  line: number; // 1-based
  column: number; // 1-based
}
export interface LocusRange {
  start: LocusPosition;
  end: LocusPosition;
}
export interface Locus {
  path?: string;
  root?: string;
  range?: LocusRange;
  symbol?: string;
  /** The pane/agent that drove this update ('editor' | 'outline' | 'dash' | …). */
  source?: string;
  /** Client instance id, for self-echo suppression across the /ws mirror. */
  origin?: string;
}

const myOrigin = Math.random().toString(36).slice(2);
let current: Locus = {};
const listeners = new Set<() => void>();

function emit(): void {
  listeners.forEach((l) => l());
}

/**
 * Publish a locus from `source`. Notifies local followers synchronously and mirrors
 * over `/ws` so dash / the agent / other windows follow.
 */
export function setLocus(locus: Locus, source: string): void {
  current = { ...locus, source, origin: myOrigin };
  emit();
  sendChannel('code', 'locus', current);
}

/** Apply a locus that arrived from the backend / another window — no re-send. */
function applyRemote(locus: Locus): void {
  current = locus;
  emit();
}

let wired = false;
function ensureWired(): void {
  if (wired) return;
  wired = true;
  subscribeChannel('code', (msg: WsMessage) => {
    if (msg.event !== 'locus') return;
    const data = msg.data as Locus | undefined;
    if (!data || data.origin === myOrigin) return; // ignore our own echo
    applyRemote(data);
  });
}

/** The current locus (empty object if none set yet). */
export function getLocus(): Locus {
  return current;
}

/** Subscribe to locus changes; returns an unsubscribe. Lazily wires the /ws mirror. */
export function subscribeLocus(listener: () => void): () => void {
  ensureWired();
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/** React hook: the current locus, re-rendering the component on every change. */
export function useLocus(): Locus {
  return useSyncExternalStore(subscribeLocus, getLocus, getLocus);
}
